### ISSUE TITLE
Calling `from` fn on an associated type that is private doesn't warn and compiler end considering the one from From trait

### ISSUE THREAD

**Issue by spastorino** [rust-lang/rust#53868]
Let say we have a main.rs with this content ...

```rust
use std::io::{self, Write};
use std::process::Command;

mod cmd;
mod error;

use cmd::Cmd;
use error::Error;

fn main() -> Result<(), io::Error> {
    let stdin = io::stdin();
    let mut stdout = io::stdout();

    loop {
        let mut line = String::new();
        print!("> ");
        stdout.flush()?;
        stdin.read_line(&mut line)?;

        match Cmd::from(&line) {
            Ok(cmd) => {
                let output = Command::new(cmd.binary).args(cmd.args).output()?;
                print!("{}", String::from_utf8_lossy(&output.stdout));
            }
            Err(Error::NoBinary) => {}
        }
    }
}
```

error.rs is ...

```rust
#[derive(Debug)]
pub enum Error {
    NoBinary,
}
```

and cmd.rs is ...

```rust
use error::Error;

// A command consists of a binary and its arguments
pub struct Cmd<'a> {
    pub binary: &'a str,
    args: Vec<&'a str>,
}

impl<'a> Cmd<'a> {
    // Extract the command and its arguments from the commandline
    fn from(line: &'a str) -> Result<Self, Error> {
        let mut parts = line.split_whitespace();
        let binary = parts.nth(0).ok_or_else(|| Error::NoBinary)?;
        let args = parts.collect();

        Ok(Cmd { binary, args })
    }
}
```

and you get the following error when trying to compile ...

```
error[E0308]: mismatched types
  --> src/main.rs:20:25
   |
20 |         match Cmd::from(&line) {
   |                         ^^^^^ expected struct `cmd::Cmd`, found reference
   |
   = note: expected type `cmd::Cmd<'_>`
              found type `&std::string::String`

error[E0308]: mismatched types
  --> src/main.rs:21:13
   |
21 |             Ok(cmd) => {
   |             ^^^^^^^ expected struct `cmd::Cmd`, found enum `std::result::Result`
   |
   = note: expected type `cmd::Cmd<'_>`
              found type `std::result::Result<_, _>`

error[E0308]: mismatched types
  --> src/main.rs:25:13
   |
25 |             Err(Error::NoBinary) => {}
   |             ^^^^^^^^^^^^^^^^^^^^ expected struct `cmd::Cmd`, found enum `std::result::Result`
   |
   = note: expected type `cmd::Cmd<'_>`
              found type `std::result::Result<_, _>`
```

The code is wrong because the from fn is not public. As soon as you mark it as pub everything works as expected.

Anyway, I'm reporting this because I'd expect a mention in the error report that the from considered by the compiler is not the one you expect.

**Comment by cuviper** [comment]
> The problem is that I'd expect a mention in the error report that the `from` considered by the compiler is not the one you're trying to call. It's just that you forgot to mark the fn as `pub`.

In this context, the compiler thought that you meant `<Cmd as From>::from`, since `trait From` is public and in the prelude.  There is a reflexive `impl From<T> for T` affecting all sized types.

**Comment by spastorino** [comment]
Yes, for this case I was suggesting that it could give you a hint saying ... did you forgot to add pub here?.

/cc @nikomatsakis given we've talked about this over Zulip.
